syntax = "proto3";

package tctrl.schema;

import "google/protobuf/struct.proto";

option java_package = "net.t3kt.tctrl.schema";
option java_outer_classname = "TctrlSchemaProto";

enum ParamType {
    OTHER = 0;
    BOOL = 3;
    STRING = 4;
    INT = 5;
    FLOAT = 6;
    IVEC = 7;
    FVEC = 8;
    MENU = 10;
    TRIGGER = 11;
}

message ParamOption {
    string key = 1;
    string label = 2;
}

message OptionList {
    string key = 1;
    string label = 2;
    repeated ParamOption option = 3;
}

message ParamPartSpec {
    string key = 1;
    string label = 2;
    google.protobuf.Value minLimit = 3;
    google.protobuf.Value maxLimit = 4;
    google.protobuf.Value minNorm = 5;
    google.protobuf.Value maxNorm = 6;
    google.protobuf.Value defaultVal = 7;
    google.protobuf.Value value = 8;
}

message ParamSpec {
    string key = 1;
    string label = 2;
    ParamType type = 3;
    string path = 4;
    string otherType = 5;
    google.protobuf.Value minLimit = 6;
    google.protobuf.Value maxLimit = 7;
    google.protobuf.Value minNorm = 8;
    google.protobuf.Value maxNorm = 9;
    google.protobuf.Value defaultVal = 10;
    google.protobuf.Value value = 11;
    int32 valueIndex = 12;
    repeated ParamPartSpec part = 13;
    string style = 14;
    string group = 15;
    repeated ParamOption option = 16;
    string optionListKey = 17;
    repeated string tag = 18;
    string help = 19;
    string offHelp = 20;
    string buttonText = 21;
    string buttonOffText = 22;
    google.protobuf.Struct properties = 1000;
}

message ModuleTypeSpec {
    string key = 1;
    string label = 2;
    repeated ParamSpec param = 7;
    repeated GroupInfo paramGroup = 9;
}

message ModuleSpec {
    string key = 1;
    string label = 2;
    string path = 3;
    string moduleType = 4;
    string group = 5;
    repeated string tag = 6;
    repeated ParamSpec param = 7;
    repeated ModuleSpec childModule = 8;
    repeated GroupInfo paramGroup = 9;
    repeated GroupInfo childGroup = 10;
}

message ConnectionInfo {
    string key = 1;
    string label = 2;
    string type = 3;
    string host = 4;
    int32 port = 5;
}

message GroupInfo {
    string key = 1;
    string label = 2;
    repeated string tag = 3;
}

message AppSchema {
    string key = 1;
    string label = 2;
    repeated string tag = 3;
    string description = 4;
    repeated ModuleSpec children = 5;
    repeated GroupInfo childGroup = 6;
    repeated OptionList optionList = 7;
    repeated ConnectionInfo connection = 8;
    repeated ModuleTypeSpec moduleType = 9;
}
